src/custom-element.js

```diff
@@ -443,45 +464,67 @@ function createBaseCustomElementClass(win) {
      *
      * @return {?Promise}
      * @final
+     * @restricted
      */
     buildInternal() {
       assertNotTemplate(this);
-      devAssert(this.isUpgraded(), 'Cannot build unupgraded element');
+      devAssert(this.implClass_, 'Cannot build unupgraded element');
       if (this.buildingPromise_) {
         return this.buildingPromise_;
       }
-      return (this.buildingPromise_ = new Promise((resolve, reject) => {
-        const impl = this.impl_;
+
+      this.onReadyStateInternal(ReadyState.BUILDING);
+
+      // Create the instance.
+      const implPromise = this.createImpl_();
+
+      // Wait for consent.
+      const consentPromise = implPromise.then(() => {
         const policyId = this.getConsentPolicy_();
         if (!policyId) {
-          resolve(impl.buildCallback());
-        } else {
-          Services.consentPolicyServiceForDocOrNull(this)
-            .then((policy) => {
-              if (!policy) {
-                return true;
-              }
-              return policy.whenPolicyUnblock(/** @type {string} */ (policyId));
-            })
-            .then((shouldUnblock) => {
-              if (shouldUnblock) {
-                resolve(impl.buildCallback());
-              } else {
-                reject(blockedByConsentError());
-              }
-            });
+          return;
         }
-      }).then(
+        return Services.consentPolicyServiceForDocOrNull(this)
+          .then((policy) => {
+            if (!policy) {
+              return true;
+            }
+            return policy.whenPolicyUnblock(policyId);
+          })
+          .then((shouldUnblock) => {
+            if (!shouldUnblock) {
+              throw blockedByConsentError();
+            }
+          });
+      });
+
+      // Build callback.
+      const buildPromise = consentPromise.then(() =>
```

Comment:
Feel free to ignore for this PR:

I think a V2 API is as good a time as any to [rename buildCallback --> build](https://github.com/ampproject/amphtml/issues/31046) 😁

Reply:
I'd rather not do it for this PR. IMHO the naming is very distinct and we can do a find-replace with some ease at any time.